### PR TITLE
[Merged by Bors] - document more of bevy_reflect

### PR DESCRIPTION
**Pull request by dataphract**
This adds documentation for:

- The trait methods of `Reflect` and its subtraits
- The `partial_eq` and `apply` functions for `Map` et al.
- `DynamicList` and `DynamicMap`
- `TypeRegistry` and related types & traits
- `GetPath`, including an explanation of path string syntax

among other things.

Still to be documented are the various macros and `bevy_reflect::serde`.

### CODE REVIEW

File: crates/bevy_reflect/src/list.rs [hunk 1]
```diff
@@ -20,25 +33,36 @@ pub trait List: Reflect {
...
 
+/// A list of reflected values.
 #[derive(Default)]
 pub struct DynamicList {
     name: String,
     values: Vec<Box<dyn Reflect>>,
 }
 
 impl DynamicList {
+    /// Returns the name of the list.
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/list.rs:44]:
```suggestion
    /// Returns the type name of the list.
```

File: crates/bevy_reflect/src/list.rs [hunk 2]
```diff
@@ -20,25 +33,36 @@ pub trait List: Reflect {
...
+    ///
+    /// The value returned by this method is the same value returned by
+    /// [`Reflect::type_name`].
     pub fn name(&self) -> &str {
         &self.name
     }
 
+    /// Sets the name of the list.
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/list.rs:52]:
```suggestion
    /// Sets the type name of the list.
```

File: crates/bevy_reflect/src/list.rs [hunk 3]
```diff
@@ -175,6 +208,12 @@ pub fn list_apply<L: List>(a: &mut L, b: &dyn Reflect) {
     }
 }
 
+/// Compares a [`List`] with a [`Reflect`] value.
+///
+/// Returns true _iff_ all of the following are true:
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/list.rs:213]:
```suggestion
/// Returns true if and only if all of the following are true:
```
I love `iff`, but it's not clear to readers without a CS background.

File: crates/bevy_reflect/src/map.rs [hunk 4]
```diff
@@ -4,23 +4,32 @@ use bevy_utils::HashMap;
 
 use crate::{serde::Serializable, Reflect, ReflectMut, ReflectRef};
 
-/// An ordered `ReflectValue->ReflectValue` mapping. `ReflectValue` `Keys` are assumed to return a
-/// non-`None` hash. Ideally the ordering is stable across runs, but this is not required.
-/// This corresponds to types like [`std::collections::HashMap`].
+/// An ordered `ReflectValue->ReflectValue` mapping.
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/map.rs:7]: Perhaps we could have a comment here explaining that the underlying types don't have to be the same?

File: crates/bevy_reflect/src/map.rs [hunk 5]
```diff
@@ -4,23 +4,32 @@ use bevy_utils::HashMap;
...
+///
+///`ReflectValue` `Keys` are assumed to return a non-`None` hash. Ideally the
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/map.rs:9]: What does "ideally" mean here? What happens if they're not?

File: crates/bevy_reflect/src/map.rs [hunk 6]
```diff
@@ -29,18 +38,28 @@ pub struct DynamicMap {
 }
 
 impl DynamicMap {
+    /// Returns the name of the map.
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/map.rs:41]:
```suggestion
    /// Returns the type name of the map.
```

File: crates/bevy_reflect/src/map.rs [hunk 7]
```diff
@@ -29,18 +38,28 @@ pub struct DynamicMap {
...
+    ///
+    /// The value returned by this method is the same value returned by
+    /// [`Reflect::type_name`].
     pub fn name(&self) -> &str {
         &self.name
     }
 
+    /// Sets the name of the map.
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/map.rs:49]:
```suggestion
    /// Sets the type name of the map.
```

File: crates/bevy_reflect/src/map.rs (hunk 5 above)
Comment (dataphract) [reply crates/bevy_reflect/src/map.rs:9]: This was language I retained from the previous docs -- my guess is that the only visible effect is unstable serialization order between runs. Since `DynamicMap` has been an aHash-based `HashMap` under the hood since its introduction, it seems evident that stable ordering is not meant to be a guarantee of the API.

TL;DR I'm happy to remove that language if it's not relevant!

File: crates/bevy_reflect/src/type_registry.rs [hunk 8]
```diff
@@ -56,53 +63,94 @@ impl TypeRegistry {
...
+    /// Takes a write lock on the underlying [`TypeRegistry`].
     pub fn write(&self) -> RwLockWriteGuard<'_, TypeRegistry> {
         self.internal.write()
     }
 }
 
+/// A record of data about a type.
+///
+/// This contains the [`TypeId`], [name], and [short name] of the type, as well
+/// as the same data for any type parameters of the type.
```
Comment (Davier) [inline crates/bevy_reflect/src/type_registry.rs:149]: >as well as the same data for any type parameters of the type.

This seems incorrect.
The `data` field contains the information about the reflected traits. For instance, if a type has `#[reflect(Component)]`, the data field will contain an entry at the key `type_id::<ReflectComponent>()` with a boxed`ReflectComponent as dyn TypeData`.

File: crates/bevy_reflect/src/map.rs (hunk 5 above)
Comment (alice-i-cecile) [reply crates/bevy_reflect/src/map.rs:12]: This makes a lot of sense to me. If we can verify that assumption, I would be very happy to add this explanation to the docs; it's useful information!

File: crates/bevy_reflect/src/reflect.rs [hunk 9]
```diff
@@ -34,25 +46,56 @@ pub enum ReflectMut<'a> {
...
+    /// [type name]: std::any::type_name
     fn type_name(&self) -> &str;
+
+    /// Returns the value as a [`&dyn Any`][std::any::Any].
     fn any(&self) -> &dyn Any;
+
+    /// Returns the value as a [`&mut dyn Any`][std::any::Any].
     fn any_mut(&mut self) -> &mut dyn Any;
     fn apply(&mut self, value: &dyn Reflect);
     fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>>;
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_reflect/src/reflect.rs:60]: These two methods were by far the hardest for me to grok when working with reflection; we should prioritize good docs for them, especially at the trait level.

[the remark this answers is not in the document]
Comment (dataphract) [reply crates/bevy_reflect/src/reflect.rs]: I think I nailed down the semantics in the most recent commit, but do feel free to correct me if I missed or misinterpreted anything!

**Comment by alice-i-cecile** [comment]
@dataphract can you resolve the comments which have been resolved for us? :)

File: crates/bevy_reflect/src/reflect.rs [hunk 10]
```diff
@@ -34,27 +46,98 @@ pub enum ReflectMut<'a> {
...
+    ///   applied to the corresponding numbered field of `self.` Fields which
+    ///   are not present in both values are ignored.
+    /// - If `T` is a value type (i.e., neither of the above), then `value` is
+    ///   downcast to `T`, cloned, and assigned to `self`.
+    ///
+    /// Note that deriving `Reflect` for collections may result in undesired
+    /// behavior, such as applying not only the elements of `value` to those of
+    /// the other but also `value`'s metadata. For collection semantics,
+    /// consider manually implementing `Reflect` and using a helper function
+    /// such as [`list_apply`] to implement this method.
```
Comment (Davier) [inline crates/bevy_reflect/src/reflect.rs:77]: I honestly don't know what happens when you try to derive Reflect on collections, I'll try it out. What metadata are you referring to?

I think we can detail the behaviour of tuples, maps and lists here too, so that external implementations have to conform with what we are doing for `Vec` and `HashMap`. (We should also change what HashMap does, but that's for another PR)

**Review by Davier — COMMENTED** [review]
One more comment, but I have to say you've done a great job!

File: crates/bevy_reflect/src/reflect.rs (hunk 10 above)
Comment (dataphract) [reply crates/bevy_reflect/src/reflect.rs:77]: By metadata I mean values that are part of the type, but aren't elements of the collection. Say a user has their own keyed `HashMap` type that they `#[derive(Reflect)]` on -- a derived implementation would ostensibly not apply only the entries, but also the hash key of `value`, which would produce undesired effects.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_reflect/src/reflect.rs [hunk 11]
```diff
@@ -59,29 +59,36 @@ pub unsafe trait Reflect: Any + Send + Sync {
...
+    /// method are as follows:
+    /// - If `T` is a [`Struct`], then the value of each named field of `value` is
+    ///   applied to the corresponding named field of `self`. Fields which are
+    ///   not present in both structs are ignored.
+    /// - If `T` is a [`TupleStruct`] or [`Tuple`], then the value of each
+    ///   numbered field is applied to the corresponding numbered field of
+    ///   `self.` Fields which are not present in both values are ignored.
+    /// - If `T` is a [`List`], then each element of `value` is applied to the
+    ///   corresponding element of `self`. The number of elements applied is
+    ///   equal to the length of the shorter list.
```
Comment (Davier) [inline crates/bevy_reflect/src/reflect.rs:72]: Up to `self.len()` items are applied, and extra items in `value` are pushed

**Review by Davier — COMMENTED** [review]
Everything looks great. The CI / ci check failure seems unrelated, but CI / check-doc should be fixed:
`error: unresolved link to `list_apply``

**Review by Davier — APPROVED** [review]
